# Recursividade
***

Uma função recursiva é uma função que se auto referência

Em uma função recursiva temos a seguinte lógica:

* Critério de parada:

```py
if (n == 1):
    return n
```

* Recursividade:

```py
return recursive_factorial(n-1)*n
```

***
#### Exemplos fatorial normal
***

In [1]:
# Definir a função de fatorial normal
def factorial(n):
    fat = 1
    for i in range(1, n+1):
        fat *= i
    return fat

In [2]:
print(factorial(5))

120


***
#### Exemplo fatorial recursivo
***

In [3]:
# Definir a função de fatorial recursiva
def recursive_factorial(n):
    if n == 1:
        return n
    return recursive_factorial(n-1)*n

In [4]:
print(recursive_factorial(5))

120


***
#### Como funciona?
***

```py
recursive_ractorial(5):
    5 == 1? Não
    return recursive_factorial(4)*5
```

```py
recursive_factorial(4):
    4 == 1? Não
    return recursive_factorial(3)*4
```

```py
recursive_factorial(3):
    3 == 1? Não
    return recursive_factorial(2)*3
```

```py
recursive_factorial(2):
    2 == 1? Não
    return recursive_factorial(1)*2
```

```py
recursive_factorial(1):
    1 == 1? Sim
    return 1
```

Pronto descobrimos o recursive_factorial(1) = 1

* Logo o recursive_factorial(2) = $1\times2$
* Logo o recursive_factorial(3) = $1\times2\times3$
* Logo o recursive_factorial(4) = $1\times2\times3\times4$
* Logo o recursive_factorial(5) = $1\times2\times3\times4\times5 = 120$